r2-core/src/main/java/com/linkedin/r2/filter/transport/ServerRetryFilter.java

```diff
@@ -74,14 +129,132 @@ public void onStreamError(Throwable ex,
+  }
+
+  private void updateRetryDecision()
+  {
+    long currentTime = _clock.currentTimeMillis();
+
+    synchronized (_lock)
+    {
+      // Check if the current interval is stale
+      if (currentTime >= _lastRollOverTime + _updateIntervalMs)
+      {
+          // Rollover stale intervals until the current interval is reached
+          for (long time = currentTime; time >= _lastRollOverTime + _updateIntervalMs; time -= _updateIntervalMs)
+          {
+            _serverRetryTracker.rollOverStats();
+          }
+
+          _doNotRetry = _serverRetryTracker.getRetryRatio() > _maxRequestRetryRatio;
+          _lastRollOverTime = currentTime;
+      }
+    }
+  }
+
+  /**
+   * Stores the number of requests categorized by number of retry attempts. It uses the information to estimate
+   * a ratio of how many requests are being retried in the cluster. The ratio is then compared with
+   * {@link ServerRetryFilter#_maxRequestRetryRatio} to make a decision on whether or not to retry in the
+   * next interval. When calculating the ratio, it looks at the last {@link ServerRetryTracker#_aggregatedIntervalNum}
+   * intervals by aggregating the recorded requests.
+   */
+  private static class ServerRetryTracker
+  {
+    private final int _retryLimit;
+    private final int _aggregatedIntervalNum;
+
+    private final LinkedList<int[]> _retryAttemptsCounter;
+    private final int[] _aggregatedRetryAttemptsCounter;
+
+    private ServerRetryTracker(int retryLimit, int aggregatedIntervalNum)
+    {
+      _retryLimit = retryLimit;
+      _aggregatedIntervalNum = aggregatedIntervalNum;
+
+      _aggregatedRetryAttemptsCounter = new int[_retryLimit + 1];
+      _retryAttemptsCounter = new LinkedList<>();
+      _retryAttemptsCounter.add(new int[_retryLimit + 1]);
+    }
+
+    public synchronized void add(int numberOfRetryAttempts)
+    {
+      if (numberOfRetryAttempts <= _retryLimit)
+      {
+        _retryAttemptsCounter.getLast()[numberOfRetryAttempts] += 1;
+      } else
+      {
+        LOG.warn("Unexpected number of retry attempts: " + numberOfRetryAttempts + ", current retry limit: " + _retryLimit);
+      }
+    }
+
+    public synchronized void rollOverStats()
+    {
+      // rollover the current interval to the aggregated counter
+      int[] intervalToAggregate = _retryAttemptsCounter.getLast();
+      for (int i = 0; i < _retryLimit; i++)
+      {
+        _aggregatedRetryAttemptsCounter[i] += intervalToAggregate[i];
+      }
+
+      if (_retryAttemptsCounter.size() > _aggregatedIntervalNum)
+      {
+        // discard the oldest interval
+        int[] intervalToDiscard = _retryAttemptsCounter.removeFirst();
+        for (int i = 0; i < _retryLimit; i++)
+        {
+          _aggregatedRetryAttemptsCounter[i] -= intervalToDiscard[i];
+        }
+      }
+
+      // append a new interval
+      _retryAttemptsCounter.addLast(new int[_retryLimit + 1]);
+    }
+
+    public double getRetryRatio()
```

Comment:
The concurrency safety of this class is a little hard to follow currently. It looks like things should be safe, because you are using this and `rollOverStats` while holding a lock, and `add` and `rollOverStats` are synchronized and should be safe to use together. It would be nice to have the safety be much clearer though.

One option would be to make this synchronized for clarity. This will introduce contention with calls to `add`, but the method is small and should execute quickly, and not very frequently. Another option would be to consolidate this into a `rolloverStatsAsNeededAndGetRetryRatio` where this class can manage the thread safety and use separate locks to guard the list and the array. Finally you could just document the current behavior with comments.

Whichever way you choose I'd suggest annotating the class with `@ThreadSafe` or `@NotThreadSafe`, and annotate the fields accessed concurrently with `@GuardedBy`.

Reply:
I rewrote the thread-safety logic in both `RetryClient` and `ServerRetryFilter`, and annotated them with `@ThreadSafe` and `@GuardedBy`. Hopefully it's cleaner this time.